### PR TITLE
Update ABCI for new PubKey type and changes to BeginBlock and InitChain

### PR DESCRIPTION
**Pull request by ebuchman**
Update to the latest ABCI with new:
- PubKey
- RequestBeginBlock
- RequestInitChain

Still a couple TODOs left

### CODE REVIEW

**Comment by ebuchman** [comment]
Yeh let's do it. Damn looks like I need to fix some tests.

File: state/execution.go [hunk 1]
```diff
@@ -240,12 +225,56 @@ func execBlockOnProxyApp(logger log.Logger, proxyAppConn proxy.AppConnConsensus,
...
+	}
+
+	byzVals := make([]abci.Evidence, len(block.Evidence.Evidence))
+	for i, ev := range block.Evidence.Evidence {
+		// We need the validator set. We already did this in validateBlock.
+		// TODO: Should we instead cache the valset in the evidence itself and add
+		// `SetValidatorSet()` and `ToABCI` methods ?
+		valset, err := LoadValidators(stateDB, ev.Height())
+		if err != nil {
+			panic(err) // shoudn't happen
```
Comment (ebuchman) [inline state/execution.go]: shouldn't

File: types/protobuf.go [hunk 2]
```diff
@@ -1,71 +1,166 @@
...
 
-func (tm2pb) Validator(val *Validator) types.Validator {
-	return types.Validator{
-		PubKey: val.PubKey.Bytes(),
-		Power:  val.VotingPower,
+func (tm2pb) PubKey(pubKey crypto.PubKey) abci.PubKey {
+	switch pk := pubKey.(type) {
+	case crypto.PubKeyEd25519:
+		return abci.PubKey{
+			Type: "ed25519",
```
Comment (ebuchman) [inline types/protobuf.go]: use new types

File: types/protobuf.go [hunk 3]
```diff
@@ -1,71 +1,166 @@
...
+			Data: pk[:],
+		}
+	case crypto.PubKeySecp256k1:
+		return abci.PubKey{
+			Type: "secp256k1",
```
Comment (ebuchman) [inline types/protobuf.go]: same

**Comment by ebuchman** [comment]
@cwgoes is this everything you need ?

**Comment by cwgoes** [comment]
> @cwgoes is this everything you need ?

Yes, I think so, updated downstream in https://github.com/cosmos/cosmos-sdk/pull/1115.

**Pull request merged**